### DISCUSSION TITLE
AutotitleConversationJob not working for Ollama

### DISCUSSION THREAD

**Discussion by mattlindsey** [AllYourBot/hostedgpt#577] [General]
Conversations with Ollama don't get their title updated from 'New chat' because this job returns before executing.  This is because Ollama does not require an API key.
If this line is removed, AutotitleConversationJob works fine.  Is the line definitely required?  Or how should we fix it?

https://github.com/AllYourBot/hostedgpt/blob/b8c9c52dd4916697cdf0332a9cd181946d629834/app/jobs/autotitle_conversation_job.rb#L10

**Comment by mattlindsey** [comment]
There's a workaround for now by putting something in `Token` in the Settings:

![image](https://github.com/user-attachments/assets/f80874d9-fc19-4c1d-a74d-321ba4482328)

**Comment by keithschacht** [comment]
@mattlindsey Hmm, I see what you're saying. The reason I added this check is before we had this job failing and it would keep retrying so I was attempting to gracefully short circuit.

It's notable that the app does not cleanly handle the case where a token is not required. In fact, I just searched the code base and found this line:
https://github.com/AllYourBot/hostedgpt/blob/da4b146171c68e3627e1f0267818f6c7e2619505/app/services/ai_backend/open_ai.rb#L20

This makes it seem like the open_ai backend should not work at all if the effective token is blank. But you're saying this LLM works with a blank token?

**Reply by mattlindsey** [reply]
Yeah. If you just download the Ollama app and run it you don't need a key I think.  If you comment out that line at least, it works fine with no key entered in Settings.

You can just do this with no key:
```
curl http://localhost:11434/v1/chat/completions \
    -H "Content-Type: application/json" \
    -d '{
        "model": "llama3.1",
        "messages": [
            {
                "role": "user",
                "content": "Hello!"
            }
        ]
    }'

**Reply by keithschacht** [reply]
Okay so it sounds like what we need to do is add official support for “no key”. Maybe for the services setting screen we need a checkbox beneath token which says “No token needed” and if the user checks that, it greys out the token input and we save this boolean state. Or if we wanted to get clever about it we could try to differentiate between a token value of `nil` and a token value of `””`. Maybe we don’t add a boolean column but we instead add a visual checkbox which sets the value of that textbox to “” whereas if the textbox is left empty it gets set to nil? That would take some testing because I don’t know how rails controllers & models currently save empty inputs. I don’t know which would be better. It may just be easier to add a new column possibly named `token_unnecessary` or something like that.

**Comment by keithschacht** [comment]
Fixed by #583

**Comment by keithschacht** [comment]
Thanks @mattlindsey !
